### PR TITLE
Ensure single sequence values are homogeneous

### PR DESCRIPTION
**Pull request by alberttorosyan**
Added object type checking in `Run.track()`. The logic is the following:
Per each run sequence there's a `SequenceInfo` in-memory object which holds values count and data type for the sequence.
When the first value is added, `SequenceInfo` is initialized. There are two possible cases:
1. The sequence is constructed from scratch, i.e. there's nothing written in sequence container yet. In this case data type is set to be equal to the first tracked value's type.
2. The sequence is continued on top of existing one. Here the data type is set to the one written in run's metadata container.  Since there might be repositories, with logged metrics generated prior to adding `dtype`, we treat missing value as if it was `dtype float`.

Values incompatible with the sequence data type are rejected and `ValueError` is raised.
There's a special handling when value type is list. In case when the first tracked value is an empty list, we cannot check inner object data type, and the sequence `dtype` is allowed to change. Once a non-empty list is passed to `Run.track()` sequence type is updated. No other object types are allowed since that, though empty lists are still valid input for that sequence.

### CODE REVIEW

**Comment by alberttorosyan** [comment]
resolves #861

File: aim/sdk/run.py [hunk 1]
```diff
@@ -371,17 +377,28 @@ def _track_impl(
...
-            max_idx = len(val_view)
-        step = step or max_idx
-
-        self.series_counters[metric.selector] = max_idx + 1
-
-        if max_idx == 0:
+        seq_info = self.sequence_info[metric.selector]
+        if not seq_info.initialized:
+            seq_info.count = len(val_view)
+            seq_info.trace_dtype = self.meta_run_tree.get(('traces', ctx.idx, name, 'dtype'), None)
```
Comment (mihran113) [inline aim/sdk/run.py:383]: `sequence_dtype` instead of `trace_dtype`?

File: aim/sdk/run.py [hunk 2]
```diff
@@ -371,17 +377,28 @@ def _track_impl(
...
+                self.meta_tree['traces_types', new_dtype, ctx.idx, name] = 1
+                seq_info.trace_dtype = self.meta_run_tree['traces', ctx.idx, name, 'dtype'] = new_dtype
+            compatible = check_types_compatibility(dtype, seq_info.trace_dtype, update_trace_dtype)
+            if not compatible:
+                raise ValueError(f'Cannot log value \'{value}\' on sequence \'{name}\'. Incompatible data types.')
+
+        step = step or seq_info.count
+
+        if seq_info.count == 0:
             # TODO [AT] check sequence is homogeneous & handle empty list case
```
Comment (mihran113) [inline aim/sdk/run.py:399]: this `TODO` can be deleted, right?

File: aim/sdk/run.py (hunk 1 above)
Comment (alberttorosyan) [reply aim/sdk/run.py:383]: 👌

File: aim/sdk/run.py (hunk 2 above)
Comment (alberttorosyan) [reply aim/sdk/run.py:399]: yeah, that's the whole point of this PR

**Pull request merged**